library/pkcs7.c

```diff
@@ -596,6 +688,37 @@ int mbedtls_pkcs7_parse_der(mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
     return ret;
 }
 
+
+static int pkcs7_signed_attrs_match(mbedtls_asn1_named_data *attrs,
+                                    const unsigned char *hash,
+                                    size_t hashlen)
+{
+    mbedtls_asn1_named_data *attr;
+    unsigned char *p;
+    size_t len;
+    int has_digest = 0;
+
+    for (attr = attrs; attr != NULL; attr = attr->next) {
+        if (MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS9_MESSAGE_DIGEST, &attr->oid) == 0) {
+            p = attr->val.p;
+            if (mbedtls_asn1_get_tag(&p, p + attr->val.len, &len,
+                                     MBEDTLS_ASN1_OCTET_STRING) != 0) {
+                return MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+            }
+            if (len != hashlen || memcmp(p, hash, hashlen) != 0) {
+                return MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+            }
+            has_digest = 1;
+        }
+    }
+
+    if (!has_digest) {
+        return MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+    }
+
+    return 0;
+}
```

Comment:
Checking that there is exactly one `MBEDTLS_OID_PKCS9_MESSAGE_DIGEST` attribute should most likely be done during parsing as well.